controllers/lease_controller.go

```diff
@@ -0,0 +1,456 @@
+		if errors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, err
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{
+			RequeueAfter: 10 * time.Second,
+		}, err
+	}
+
+	logger := lc.logger.WithValues("etcd", kutil.Key(etcd.Namespace, etcd.Name).String())
+
+	// Get or Create delta_snapshot_revisions lease object that will keep track of delta snapshot revisions based on which
+	// compaction job will be scheduled
+	nsName := types.NamespacedName{
+		Name:      getDeltaLease(etcd),
+		Namespace: etcd.Namespace,
+	}
+
+	deltaLease := &coordinationv1.Lease{}
+	err := lc.Get(ctx, nsName, deltaLease)
+	if err != nil {
+		logger.Info("Couldn't fetch delta snap lease because: " + err.Error())
+
+		return ctrl.Result{
+			RequeueAfter: 10 * time.Second,
+		}, err
+	}
+
+	// Run compaction job
+	if etcd.Spec.Backup.Store != nil {
+		hi, err := strconv.ParseInt(*deltaLease.Spec.HolderIdentity, 10, 32)
+		if err != nil {
+			logger.Error(err, "Can't convert holder identity of revision lease to integer")
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, err
+		}
+		// Reconcile job only when current revision is more than 1 million (this vale is configurable through `events-threshold` option in druid) than
+		// the last snapshot (full/compact) revision
+		if hi >= lc.config.EventsThreshold {
+			j, err := lc.reconcileJob(ctx, logger, etcd)
+			if err != nil {
+				return ctrl.Result{
+					RequeueAfter: 5 * time.Second,
+				}, err
+			}
+			logger.Info("Current compaction job is: " + j.Name)
+		}
+	}
+
+	return ctrl.Result{
+		Requeue: false,
+	}, nil
+}
+
+func (lc *LeaseController) reconcileJob(ctx context.Context, logger logr.Logger, etcd *druidv1alpha1.Etcd) (*batchv1.Job, error) {
+	logger.Info("Reconcile etcd compaction job")
+
+	// First check if a job is already running
+	job := &batchv1.Job{}
+	err := lc.Get(ctx, types.NamespacedName{Name: getJobName(etcd), Namespace: etcd.Namespace}, job)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Required job doesn't exist. Create new
+			job, err = lc.CreateCompactJob(ctx, logger, etcd)
+			if err != nil {
+				logger.Error(err, "error during compaction job creation")
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	}
+
+	// Delete job and reque if the job failed
+	if job.Status.Failed > 0 {
+		err = lc.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationForeground))
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("job status failed")
+	}
+
+	// Delete job and return if the job succeeded
+	if job.Status.Succeeded > 0 {
+		err = lc.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationForeground))
+		if err != nil {
+			return nil, err
+		}
+		return job, nil
+	}
```

Comment:
Should the controller not requeue after a successful job has been deleted? Won't it be possible that we find the previously run job, delete that, and wait for the next delta snap lease update to create a new job?